### PR TITLE
[Backport release-24.05] keycloak: 24.0.5 -> 25.0.6

### PR DESCRIPTION
**Pull request by NickCao**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/servers/keycloak/default.nix [hunk 1]
```diff
@@ -7,7 +7,7 @@
 , callPackage
 , confFile ? null
 , plugins ? [ ]
-, extraFeatures ? [ ]
+, extraFeatures ? [ "hostname:v1" ]
```
Comment (teutat3s) [inline pkgs/servers/keycloak/default.nix:10]: Looks like this needs to be 
```suggestion
, extraFeatures ? [ "hostname-v1" ]
```

**Review by teutat3s — COMMENTED** [review]
It's probably failing because the feature name is not valid. I also saw references to `hostname:v1` in the docs, but it's seems the feature name is `hostname-v1`.
Source: https://www.keycloak.org/server/features

File: pkgs/servers/keycloak/default.nix (hunk 1 above)
Comment (NickCao) [reply pkgs/servers/keycloak/default.nix:10]: This fails even earlier, at the build time.

**Comment by mweinelt** [comment]
Any particular reason for this backport?

File: pkgs/servers/keycloak/default.nix (hunk 1 above)
Comment (teutat3s) [reply pkgs/servers/keycloak/default.nix:10]: It's failing [here](https://github.com/keycloak/keycloak/blob/25.0.1/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java#L310), might need to set `hostname-port`?

Comment (NickCao) [reply pkgs/servers/keycloak/default.nix:10]: Doesn't work either.

**Comment by teutat3s** [comment]
Some additional observations that caught my eye while trying to debug this. I could not spot clues to a solution yet.

```
❯ nix build --print-build-logs -f . nixosTests.keycloak.postgres.driverInteractive
...
WARN  [org.key.com.Profile] (main) Deprecated features enabled: hostname:v1
WARN  [org.key.qua.run.cli.Picocli] (main) The following run time non-cli options were found, but will be ignored during build time:
  kc.db-url-host, kc.db-url-database, kc.db-url-port, kc.db-url-properties, kc.db-username, kc.db-password, kc.hostname,
  kc.hostname-strict-backchannel, kc.http-host, kc.http-port, kc.https-port, kc.https-certificate-file, kc.https-certificate-key-file, kc.proxy
...
```
and
```
$ kc.sh --config-file /nix/store/j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf show-config
Current Mode: production
Current Configuration:
	kc.config.built =  true (SysPropConfigSource)
	kc.config.file =  /nix/store/j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf (SysPropConfigSource)
	kc.db =  postgres (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-password =  ******* (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-database =  keycloak (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-host =  localhost (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-port =  5432 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-url-properties =   (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.db-username =  keycloak (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.hostname =  acme.test (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-host =  0.0.0.0 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-port =  80 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.http-relative-path =  / (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-certificate-file =  /run/keycloak/ssl/ssl_cert (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-certificate-key-file =  /run/keycloak/ssl/ssl_key (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.https-port =  443 (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.log-console-output =  default (classpath keycloak.conf)
	kc.proxy =  none (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.spi-hostname-v2-hostname =  acme.test (j05dgml8criznlavvxz9yd9zp7rirshy-keycloak.conf)
	kc.version =  25.0.1 (SysPropConfigSource)
```

**Comment by NickCao** [comment]
In an unexpected turn of events, I found out why this was not working in https://github.com/NixOS/nixpkgs/pull/346439

**Comment by teutat3s** [comment]
Result of `nixpkgs-review pr 323480` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 tests built:</summary>
  <ul>
    <li>nixosTests.keycloak.mariadb</li>
    <li>nixosTests.keycloak.mysql</li>
    <li>nixosTests.keycloak.postgres</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>pkgs.keycloak</li>
  </ul>
</details>

**Review by teutat3s — APPROVED** [review]

**Comment by teutat3s** [comment]
@ofborg test keycloak

**Comment by NickCao** [comment]
Let's add the commit raising test vm memory size, should help with the flaky tests.

**Review by teutat3s — APPROVED** [review]

**Comment by NickCao** [comment]
So they only support one major version: https://endoflife.date/keycloak

We might as well leave it as is. Version 25 is no better than 24 in terms of support now.

**Comment by dkowis** [comment]
https://www.keycloak.org/2024/10/release-updates

Just FYI, keycloak is changing the way they do releases so that it's less disruptive, and hopefully this will be easier to deal with.

**Pull request merged**